### DISCUSSION TITLE
Importing a custom worker

### DISCUSSION THREAD

**Discussion by jesse-mm** [serwist/serwist#24] [Q&A]
Hi @DuCanhGH!

Thanks for your hard work on next-pwa fork and now on serwist. I've migrated from next-pwa to serwist but cannot find documentation around importing a custom worker. 

In the next-pwa fork this could be done by adding a `worker/index.js` ([ref](https://ducanh-next-pwa.vercel.app/docs/next-pwa/custom-worker)). How would I do this using serwist, I'm now using `importScripts` functionality in `installSerwist` but these files are not processed by webpack.

Would be great if you have a solution for this :)

**Comment by jesse-mm** [comment]
Hmm I see https://github.com/serwist/serwist/pull/1

> Custom workers have been removed. You should use swSrc.

Does this mean I can import my custom worker logic in `sw.ts` ?

**Comment by DuCanhGH** [comment]
> Does this mean I can import my custom worker logic in sw.ts ?

Yeah, damn right you can.
Most of the custom worker's use cases are covered by simply... writing that code in `swSrc`. However, if that does not work, you can do this:
- For Next.js/Webpack, you can import these custom worker code through static `import` declarations like how you'd usually do it in your app. If that's not your cup of tea, however, your only option is to manually bundle the files you want to use `importScripts` with, and figure out a way to detect them in your service worker. To do this, you can create a Webpack plugin that bundles all the files in a specific folder and returns a list of bundled files, which you can send to your service worker through `webpack.DefinePlugin`. This will work because we use the same Webpack instance as yours to bundle the service worker. I may provide a full example if you do need that, since configuring Webpack... simply isn't everyone's favourite thing to do.
- For Vite, you can use static `import` declarations, same as Next.js/Webpack. Moreover, you are also allowed to import the workers as URLs:
```ts
import type { PrecacheEntry } from "@serwist/precaching";
import { installSerwist } from "@serwist/sw";
import { defaultCache } from "@serwist/vite/worker";

import customWorker from "./custom-worker.js?worker&url";

declare const self: ServiceWorkerGlobalScope & {
  // Change this attribute's name to your `injectionPoint`.
  __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
};

installSerwist({
  importScripts: [customWorker],
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});
```

What Vite does is bundling the referenced file and replacing the import in the service worker with the URL to this bundled file.

**Reply by jesse-mm** [reply]
Thanks for the swift reply this was really insightful and clears up things! I'll fiddle with it 🙌 !
